rpc/dataconn/dataconn.go

```diff
@@ -0,0 +1,600 @@
+outer:
+	for ctx.Err() == nil {
+
+		// fork off the initial Serve-call context for each connection
+		ctx := context.WithValue(ctx, contextKeyReset, nil)
+
+		accepted := make(chan struct{})
+		go func() {
+			select {
+			case <-ctx.Done():
+				if err := l.Close(); err != nil {
+					getLog(ctx).WithError(err).Error("error closing listener after context error")
+				}
+			case <-accepted:
+			}
+		}()
+		wire, err := l.Accept()
+		close(accepted)
+		if err != nil {
+			getLog(ctx).WithError(err).Error("accept error")
+			continue
+		}
+		if s.wireInterceptor != nil {
+			ctx, wire = s.wireInterceptor(ctx, wire) // SHADOWING
+		}
+		clientError := func(msgToClient string) {
+			// the client may be malicious, use a hard deadline through timed Close
+			// (we don't use SetDeadline because that may be the cause for the call)
+			dl := time.NewTimer(s.config.HeaderTimeout)	
+			messageSent := make(chan struct{})
+			go func() {
+				defer close(messageSent)
+				var res reqHeader
+				res.HandlerErr = msgToClient
+				if err := res.marshalToWire(wire); err != nil {
+					getLog(ctx).WithError(err). // not an error, could be malicious
+									Debug("response after failed header unmarshal failed")
+				}
+			}()
+			go func() {
+				select {
+				case <-messageSent:
+					dl.Stop()
+				case <-dl.C:
+					getLog(ctx).Debug("sending client error timed out")
+				}
+				getLog(ctx).Debug("closing client connection")
+				if err := wire.Close(); err != nil {
+					getLog(ctx).WithError(err).
+						Error("error closing client conection after previous error")
+				}
+			}()
+		}
+
+		// header timeout, reset before call to handler
+		headerDL := timeoutToDeadline(s.config.HeaderTimeout)
+		if err := wire.SetDeadline(headerDL); err != nil {
+			getLog(ctx).WithError(err).
+				Error("cannot set connection deadline")
+			clientError("internal server error")
+			continue
+		}
+
+		var hdr reqHeader
+		if err := hdr.unmarshalFromWire(wire, s.config.Shared.MaxHeaderLen); err != nil {
+			getLog(ctx).WithError(err).Debug("unmarshaling header failed") // not an error, could be malicious
+			clientError(fmt.Sprintf("unmarshalling header failed: %s", err))
+			continue
+		}
+
+		if hdr.ProtoLen > s.config.Shared.MaxProtoLen {
+			getLog(ctx).WithError(err).
+				WithField("max_proto_length", s.config.Shared.MaxProtoLen).
+				Debug("request proto exceeds max length")
+			clientError(fmt.Sprintf("protolen exceeds configured maximum"))
+			continue
+		}
+
+		protoBuf := make([]byte, hdr.ProtoLen)
+		if _, err := io.ReadFull(wire, protoBuf); err != nil {
+			clientError(fmt.Sprintf("reading protobuf failed: %s", err))
+			continue
+		}
+
+		// disable header timeouts (absolute deadlines) while in handler
+		if err := wire.SetDeadline(time.Time{}); err != nil {
+			getLog(ctx).WithError(err).Error("cannot unset connection deadline")
+		}
+
+		getLog(ctx).Debug("pre handler")
+
+		// FIXME wrap conn into an idle-timeout conn
```

Comment:
no go: try setting up a test scenario for stalled timeouts.
then implement timeout handling.
consider re-using keepaliveio (still refd by tag `grpc_dataconn_rpc_prerebase` on my local machine)